genesis-tools/lockup-contract/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+    }
+
+    /// Get the key type of the given key.
+    fn get_key_type(&self, key: &PublicKey) -> Option<KeyType> {
+        self.keys
+            .iter()
+            .find_map(|(key_type, keys)| if keys.contains(key) { Some(*key_type) } else { None })
+    }
+
+    /// Adds the key both to the internal collection and through promise without
+    /// checking the permissions of the signer.
+    fn add_key_no_check(&mut self, key_type: KeyType, key: PublicKey) {
+        self.keys.entry(key_type).or_default().insert(key.clone());
+        match key_type {
+            KeyType::Full => {
+                let account_id = env::current_account_id();
+                Promise::new(account_id).add_full_access_key(key);
+            }
+            key_type @ _ => {
+                let account_id = env::current_account_id();
+                Promise::new(account_id.clone()).add_access_key(
+                    key,
+                    0,
+                    account_id,
+                    key_type.allowed_methods().to_vec(),
+                );
+            }
+        }
+    }
+
+    /// Add the new access key. The signer access key should have permission to do it.
+    pub fn add_key(&mut self, key_type: KeyType, key: PublicKey) {
+        self.signer_key_type().check_can_add_remove(&key_type);
+        self.add_key_no_check(key_type, key);
+    }
+
+    /// Remove a known access key. The signer access key should have enough permission to do it.
+    pub fn remove_key(&mut self, key: PublicKey) {
+        let key_type_to_remove = self.get_key_type(&key).expect("Cannot remove unknown key");
+        self.signer_key_type().check_can_add_remove(&key_type_to_remove);
+        Promise::new(env::current_account_id()).delete_key(key);
+    }
+
+    /// Create a staking transaction on behalf of this account.
+    pub fn stake(&self, amount: u128, public_key: PublicKey) {
+        assert!(!self.permanently_unstaked, "The account was permanently unstaked.");
+        assert!(amount <= env::account_balance(), "Not enough balance to stake.");
+        Promise::new(env::current_account_id()).stake(amount, public_key);
+    }
+
+    /// Get the amount of liquid tokens that this account has.
+    #[cfg(not(feature = "vesting"))]
+    pub fn get_liquid(&self) -> u128 {
+        if env::block_timestamp() >= self.lockup_timestamp {
+            env::account_balance()
+        } else {
+            env::account_balance() - self.lockup_amount
+        }
+    }
+
+    /// Get the amount of tokens that were vested.
+    #[cfg(feature = "vesting")]
+    pub fn get_vested(&self) -> u128 {
+        let block_timestamp = env::block_timestamp();
+        if block_timestamp <= self.vesting_cliff_timestamp {
+            0
+        } else if block_timestamp >= self.vesting_end_timestamp {
+            self.lockup_amount
+        } else {
+            (block_timestamp - self.vesting_start_timestamp) as u128 * self.lockup_amount
```

Comment:
Might overflow here, if timestamps are large and lockup is large. Ideally we want u256

Reply:
Fixed.

Reply:
`u256` is not available in `std`, so I worked it around.